REVIEW_COPY.md

```diff
@@ -0,0 +1,312 @@
+Apart from their coordinates in the DOM, our beacons will also need to describe their location in the outline. A beacon can define its location in relative to another node in the outline.
+
+```elm
+type CandidatePosition
+    = Before String
+    | After String
+    | PrependedIn String
+    | AppendedIn String
+```
+
+We'll create a JSON encoder for this type so we can tag each beacon element with a data attribute containing its position in the outline. We'll then set up our JavaScript to find all elements with such a data attribute in the DOM and feed their coordinates back to us on each drag event. That will allow us to define a type for drag events containing the positions of our beacons on the screen.
+
+```elm
+type alias DragMsg =
+    { cursor : Coords
+    , beacons : List Beacon
+    }
+
+
+type alias Beacon =
+    ( CandidatePosition, Rect )
+```
+
+Remember the `closestRect` function from the polygon example? It's what we need to find the `CandidatePosition` closest to the cursor! Once we know which candidate position is closest, all we need is a function to that moves a node to its new position in an outline:
+
+```elm
+moveNodeInOutline : String -> CandidatePosition -> List OutlineNode -> List OutlineNode
+moveNodeInOutline = Debug.todo "See full source code for implementation!"
+```
+
+It's a tricky function to write, but it doesn't have much to do with drag & drop and so I'm skipping the implementation here. I include a solution with the [outline editor source code]. If you're interested in some thoughts on how to approach data transformations like these, check out this earlier [post on conversion functions], which includes an example of a similar tree manipulation problem.
+
+## Necessary JavaScript
+
+I promised I'd get back at the JavaScript required to make these examples work. All three examples use the same JavaScript code, because it turns out they have the same needs. In every example there's one or more Html elements on the page that we need to track the position and dimensions of as a drag interaction takes place. What our JavaScript code needs to do is generate events when the mouse gets pressed, moved, and released, and bundle with those events the positions of all elements we want to track. We identify those elements by giving them a data attribute with their 'beacon ID'.
+
+There's tons of ways to write this code and I don't believe mine is particularly insightful, so I'll not reprint it here. The [draggable.js source code] for these examples is available though in case you're interested.
+
+## Conclusion
+
+When we need to perform a complicated task, its natural to start by looking for a library to do the heavy lifting for us. Unfortunately I've not found this approach fruitful in building drag & drop interfaces. I believe this is because the draggables and dropzones that are central to the libraries I've tried were not a great fit for the experiences I tried to build, of which I gave three examples in this post.
```

Comment:
If you were interested in adding additional content, I think it would be natural here to comment on the somewhat hidden cost of using libraries, namely working with the abstractions they've created. If those abstractions are a good fit to your domain, then no big deal, but if not then you wind up fighting those abstractions over & over.

Reply:
Yes! I'm a big believer off the statement: no abstraction is better than a bad abstraction. concept. I'd love to write more than that.

With draggables & dropzones in this post basically being a single example of that statement, I don't feel I've 'earned' the leap to this more abstract claim yet. I'd love to get back to this broader point after reading/writing one or two more blog posts illustrating the cost of bad abstractions, to have a couple of different examples to point to as supporting this claim.